metaflow/plugins/cards/card_datastore.py

```diff
@@ -153,25 +173,68 @@ def card_info_from_path(path):
+        # HACK : FIXME (LATER) : Fix the duplication of below block in a few months.
+        # Check file blame to understand the age of this HACK.
+
+        # This function will end up saving cards at two locations.
+        # Thereby doubling the number of cards. (Which is a temporary fix)
+        # Why do this ? :
+        # When cards were introduced there was an assumption made about task-ids being unique.
+        # This assumption was incorrect.
+        # Only the pathspec needs to be unique but there is no such guarantees about task-ids.
+        # When task-ids are non-unique, card read would result in finding incorrect cards.
+        # This happens because cards were stored based on task-ids.
+        # If we immediately switch from storing based on task-ids to a step-name abstraction folder,
+        # then card reading will crash for many users.
+        # It would especially happen for users who are accessing cards created by a newer
+        # MF client from an older version of MF client.
+        # It will also easily end up breaking the metaflow-ui (which maybe using a client from an older version).
+        # Hence we are writing cards to both paths so that we can introduce breaking changes later in the future.
+        card_path_with_steps = self.get_card_location(
+            self._get_write_path(), card_file_name, card_html, card_id=card_id
         )
-        self._backend.save_bytes(
-            [(card_path, BytesIO(bytes(card_html, "utf-8")))], overwrite=overwrite
+        card_path_without_steps = self.get_card_location(
+            self._get_read_path(with_steps=False),
+            card_file_name,
+            card_html,
+            card_id=card_id,
         )
-        return self.card_info_from_path(card_path)
+        if hasattr(metaflow_config, "_HACK_SKIP_CARD_DUALWRITE"):
+            self._backend.save_bytes(
+                [(card_path_with_steps, BytesIO(bytes(card_html, "utf-8")))],
+                overwrite=overwrite,
+            )
+        else:
+            for cp in [card_path_with_steps, card_path_without_steps]:
+                self._backend.save_bytes(
+                    [(cp, BytesIO(bytes(card_html, "utf-8")))], overwrite=overwrite
+                )
+
+        return self.card_info_from_path(card_path_with_steps)
 
     def _list_card_paths(self, card_type=None, card_hash=None, card_id=None):
-        card_path = self._get_card_path()
-
-        card_paths = self._backend.list_content([card_path])
-        if len(card_paths) == 0:
-            # If there are no files found on the Path then raise an error of
-            raise CardNotPresentException(
-                self._pathspec,
-                card_hash=card_hash,
-                card_type=card_type,
+        # Check for new cards first
+        card_paths = []
+        card_paths_with_steps = self._backend.list_content(
+            [self._get_read_path(with_steps=True)]
+        )
+
+        if len(card_paths_with_steps) == 0:
```

Comment:
This will "penalize" tasks that don't have cards at all (since you look in two locations). I think this is OK given that it is most likely not something you should be doing. The only possible issue is the UI which may take a tad longer to determine that there are no cards. I don't believe this is an issue either. I don't have a good solution to have a penalty free approach in all cases.

Reply:
This exception is how the user is notified that there are no cards when accessing things from the cli. Card tests also rely on these exceptions to make decisions if tests passed or failed. 

Removing this will create a lot bigger refactor , I think we can leave it here for now :)

Reply:
Sorry, wasn't suggesting to not return the exception, was just thinking if we could do so more cheaply but I think this is fine.